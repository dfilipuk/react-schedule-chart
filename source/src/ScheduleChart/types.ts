enum TimelineGranularity {
  Daily,
  Weekly,
}

type BaseStatus = string | number | symbol;

type Task<TStatus extends BaseStatus> = {
  status: TStatus;
  startDate: Date;
  endDate: Date;
};

type TaskExtended<TStatus extends BaseStatus> = {
  index: number;
  status: TStatus;
  startDate: Date;
  endDate: Date;
  relativeStartDate: number;
  relativeEndDate: number;
};

type Activity<TStatus extends BaseStatus> = {
  id: number;
  title: string;
  tasks: Task<TStatus>[];
};

type ActivityExtended<TStatus extends BaseStatus> = {
  id: number;
  title: string;
  tasks: Task<TStatus>[];
  latestCompletionDate: Date | null;
};

type ScheduleCheckpoint<TStatus extends BaseStatus> = {
  relativeDate: number;
  activeTasks: TaskExtended<TStatus>[];
  completedTasks: TaskExtended<TStatus>[];
};

type ScheduleAction<TStatus extends BaseStatus> = {
  status: TStatus;
  relativeStartDate: number;
  relativeEndDate: number;
  checkpoints: ScheduleCheckpoint<TStatus>[];
};

type Colors<TStatus extends BaseStatus> = {
  [key in TStatus]: { primary: string; secondary: string };
};

type Configuration<TStatus extends BaseStatus> = {
  colors: Colors<TStatus>;
  isCompleted: IsCompletedFunction;
  getMostRelevantStatus: GetMostRelevantStatusFunction<TStatus>;
};

type TimelineSettings = {
  granularity: TimelineGranularity;
  startDate: Date;
  endDate: Date;
  durationInDays: number;
  durationInWeeks: number;
  relativeCurrentDate: number | null;
  relativeCurrentWeek: number | null;
};

// https://stackoverflow.com/questions/56505560/how-to-fix-ts2322-could-be-instantiated-with-a-different-subtype-of-constraint
type IsCompletedFunction = (status: any) => boolean;
type GetMostRelevantStatusFunction<TStatus extends BaseStatus> = (statuses: any[]) => TStatus;

export { TimelineGranularity };
export type {
  BaseStatus,
  Task,
  TaskExtended,
  Activity,
  ActivityExtended,
  ScheduleCheckpoint,
  ScheduleAction,
  Colors,
  Configuration,
  TimelineSettings,
  IsCompletedFunction,
  GetMostRelevantStatusFunction,
};
