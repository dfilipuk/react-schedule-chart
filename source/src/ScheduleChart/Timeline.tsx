import React, { useMemo } from 'react';
import styled from 'styled-components';

import { TimelineBar } from './TimelineBar';
import { TimelineMarker } from './TimelineMarker';
import { BaseStatus, Configuration, Task, TimelineSettings } from './types';
import { calculateSchedule } from './utils';

const Container = styled.div<{ daysCount: number }>`
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(${(props) => props.daysCount}, 1fr);
`;

const CurrentDayMarker = styled.div<{ position: number }>`
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: ${(props) => props.position} / span 1;
  border-right: 1px dashed #606d7f;
`;

type Props<TStatus extends BaseStatus> = {
  className?: string;
  configuration: Configuration<TStatus>;
  timelineSettings: TimelineSettings;
  tasks: Task<TStatus>[];
  dataTestId?: string;
};

const Timeline = React.memo(<TStatus extends BaseStatus>(props: Props<TStatus>) => {
  const { className, configuration, timelineSettings, tasks: inputTasks, dataTestId } = props;
  const { colors, isCompleted, getMostRelevantStatus } = configuration;
  const { startDate, relativeCurrentDate, durationInDays } = timelineSettings;

  const scheduleActions = useMemo(
    () => calculateSchedule(inputTasks, startDate, isCompleted, getMostRelevantStatus),
    [inputTasks, startDate]
  );

  return (
    <Container data-testid={dataTestId} className={className} daysCount={durationInDays}>
      {scheduleActions.map(({ status, relativeStartDate, relativeEndDate, checkpoints }) => {
        const singleActiveTask =
          checkpoints.length === 1 &&
          checkpoints[0].activeTasks.length === 1 &&
          checkpoints[0].completedTasks.length === 0;

        return (
          <React.Fragment key={`${relativeStartDate}-${relativeEndDate}`}>
            <TimelineBar
              startDay={relativeStartDate}
              endDay={relativeEndDate}
              color={colors[status].primary}
              task={singleActiveTask ? checkpoints[0].activeTasks[0] : undefined}
            />

            {!singleActiveTask &&
              checkpoints.map((checkpoint, _, arr) => (
                <TimelineMarker
                  key={checkpoint.relativeDate}
                  checkpoint={checkpoint}
                  singleOnTimeline={arr.length === 1}
                  color={colors[status].secondary}
                />
              ))}
          </React.Fragment>
        );
      })}

      {relativeCurrentDate && <CurrentDayMarker position={relativeCurrentDate} />}
    </Container>
  );
});

export { Timeline };
