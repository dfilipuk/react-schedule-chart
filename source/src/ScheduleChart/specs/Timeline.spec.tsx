import 'jest-styled-components';

import { fireEvent, render, waitFor } from '@testing-library/react';

import { Timeline } from '../Timeline';
import { Task, TimelineGranularity, TimelineSettings } from '../types';
import * as utils from '../utils';
import { CONFIGURATION, TaskStatus } from './config';

const DEFAULT_SETTINGS: TimelineSettings = {
  granularity: TimelineGranularity.Weekly,
  durationInDays: 14,
  durationInWeeks: 2,
  startDate: new Date(2021, 5, 1),
  endDate: new Date(2021, 5, 14),
  relativeCurrentDate: null,
  relativeCurrentWeek: 2,
};

const renderTimeline = (tasks: Task<TaskStatus>[], settings: TimelineSettings = DEFAULT_SETTINGS) =>
  render(<Timeline tasks={tasks} timelineSettings={settings} configuration={CONFIGURATION} />);

describe('Timeline Component Snapshot', () => {
  it('should render current day marker if current day within timeline', () => {
    const settings: TimelineSettings = {
      ...DEFAULT_SETTINGS,
      relativeCurrentDate: 9,
    };

    const { container } = renderTimeline([], settings);

    expect(container).toMatchSnapshot();
  });

  it('should not render current day marker if current day not within timeline', () => {
    const settings: TimelineSettings = {
      ...DEFAULT_SETTINGS,
      relativeCurrentDate: null,
    };

    const { container } = renderTimeline([], settings);

    expect(container).toMatchSnapshot();
  });

  it('should render single active multiday task', () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.Open,
        startDate: new Date(2021, 5, 3),
        endDate: new Date(2021, 5, 10),
      },
    ];

    const { container } = renderTimeline(tasks);

    expect(container).toMatchSnapshot();
  });

  it('should render single active one day task', () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 5, 7),
        endDate: new Date(2021, 5, 7),
      },
    ];

    const { container } = renderTimeline(tasks);

    expect(container).toMatchSnapshot();
  });

  it('should render single completed multiday task', () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 10),
        endDate: new Date(2021, 5, 14),
      },
    ];

    const { container } = renderTimeline(tasks);

    expect(container).toMatchSnapshot();
  });

  it('should render single completed one day task', () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 1),
        endDate: new Date(2021, 5, 1),
      },
    ];

    const { container } = renderTimeline(tasks);

    expect(container).toMatchSnapshot();
  });

  it('should render several not intersecting tasks', () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 1),
        endDate: new Date(2021, 5, 3),
      },
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 5, 4),
        endDate: new Date(2021, 5, 7),
      },
      {
        status: TaskStatus.Open,
        startDate: new Date(2021, 5, 10),
        endDate: new Date(2021, 5, 12),
      },
    ];

    const { container } = renderTimeline(tasks);

    expect(container).toMatchSnapshot();
  });

  it('should render single group of several intersecting tasks', () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 5, 1),
        endDate: new Date(2021, 5, 14),
      },
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 1),
        endDate: new Date(2021, 5, 3),
      },
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 5, 4),
        endDate: new Date(2021, 5, 7),
      },
      {
        status: TaskStatus.Open,
        startDate: new Date(2021, 5, 10),
        endDate: new Date(2021, 5, 12),
      },
    ];

    const { container } = renderTimeline(tasks);

    expect(container).toMatchSnapshot();
  });

  it('should render several groups of several intersecting tasks', () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 1),
        endDate: new Date(2021, 5, 3),
      },
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 5, 2),
        endDate: new Date(2021, 5, 5),
      },
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 7),
        endDate: new Date(2021, 5, 11),
      },
      {
        status: TaskStatus.Open,
        startDate: new Date(2021, 5, 9),
        endDate: new Date(2021, 5, 12),
      },
    ];

    const { container } = renderTimeline(tasks);

    expect(container).toMatchSnapshot();
  });

  it('should render several tasks which start on single day', () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 5, 5),
        endDate: new Date(2021, 5, 10),
      },
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 5, 5),
        endDate: new Date(2021, 5, 7),
      },
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 5, 5),
        endDate: new Date(2021, 5, 12),
      },
    ];

    const { container } = renderTimeline(tasks);

    expect(container).toMatchSnapshot();
  });

  it('should render several tasks which complete on single day', () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 5),
        endDate: new Date(2021, 5, 12),
      },
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 7),
        endDate: new Date(2021, 5, 12),
      },
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 10),
        endDate: new Date(2021, 5, 12),
      },
    ];

    const { container } = renderTimeline(tasks);

    expect(container).toMatchSnapshot();
  });

  it('should render several tasks which start and complete on single day', () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 3),
        endDate: new Date(2021, 5, 8),
      },
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 5, 5),
        endDate: new Date(2021, 5, 8),
      },
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 5, 8),
        endDate: new Date(2021, 5, 8),
      },
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 5, 8),
        endDate: new Date(2021, 5, 11),
      },
    ];

    const { container } = renderTimeline(tasks);

    expect(container).toMatchSnapshot();
  });
});

describe('Timeline Component', () => {
  beforeEach(() => jest.restoreAllMocks());

  it('should render tooltip with numbered dates on hover of task end marker if it is not single marker on timeline bar', async () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 3, 2),
        endDate: new Date(2021, 3, 11),
      },
      {
        status: TaskStatus.InProgress,
        startDate: new Date(2021, 3, 4),
        endDate: new Date(2021, 3, 8),
      },
    ];
    const formatTaskDuration = jest
      .spyOn(utils, 'formatTaskDuration')
      .mockImplementation(() => 'date');

    const { getByTestId, getByText, queryByText } = renderTimeline(tasks);

    expect(queryByText('date')).not.toBeInTheDocument();

    fireEvent.mouseEnter(getByTestId('timeline-end-marker'));
    await waitFor(() => expect(getByText('date')).toBeInTheDocument());
    expect(formatTaskDuration).toHaveBeenCalledWith(tasks[0].startDate, tasks[0].endDate, 1);
  });

  it('should render tooltip with not numbered dates on hover of task end marker if it is single marker on timeline bar', async () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.Closed,
        startDate: new Date(2021, 3, 2),
        endDate: new Date(2021, 3, 11),
      },
    ];
    const formatTaskDuration = jest
      .spyOn(utils, 'formatTaskDuration')
      .mockImplementation(() => 'date');

    const { getByTestId, getByText, queryByText } = renderTimeline(tasks);

    expect(queryByText('date')).not.toBeInTheDocument();

    fireEvent.mouseEnter(getByTestId('timeline-end-marker'));
    await waitFor(() => expect(getByText('date')).toBeInTheDocument());
    expect(formatTaskDuration).toHaveBeenCalledWith(tasks[0].startDate, tasks[0].endDate, null);
  });

  it('should render tooltip on timeline bar hover if it contains single active task', async () => {
    const tasks: Task<TaskStatus>[] = [
      {
        status: TaskStatus.Open,
        startDate: new Date(2021, 3, 2),
        endDate: new Date(2021, 3, 11),
      },
    ];
    jest.spyOn(utils, 'formatTaskDuration').mockImplementation(() => 'date');

    const { getByTestId, getByText, queryByText } = renderTimeline(tasks);

    expect(queryByText('date')).not.toBeInTheDocument();

    fireEvent.mouseEnter(getByTestId('timeline-bar'));
    await waitFor(() => expect(getByText('date')).toBeInTheDocument());
  });

  test.each([
    {
      tasks: [
        {
          status: TaskStatus.InProgress,
          startDate: new Date(2021, 5, 3),
          endDate: new Date(2021, 5, 10),
        },
        {
          status: TaskStatus.InProgress,
          startDate: new Date(2021, 5, 5),
          endDate: new Date(2021, 5, 7),
        },
      ],
    },
    {
      tasks: [
        {
          status: TaskStatus.InProgress,
          startDate: new Date(2021, 5, 3),
          endDate: new Date(2021, 5, 10),
        },
        {
          status: TaskStatus.InProgress,
          startDate: new Date(2021, 5, 3),
          endDate: new Date(2021, 5, 7),
        },
      ],
    },
    {
      tasks: [
        {
          status: TaskStatus.Closed,
          startDate: new Date(2021, 5, 3),
          endDate: new Date(2021, 5, 7),
        },
        {
          status: TaskStatus.InProgress,
          startDate: new Date(2021, 5, 7),
          endDate: new Date(2021, 5, 10),
        },
      ],
    },
  ])(
    "should not render tooltip on timeline bar hover if it doesn't contain single active task: %o",
    async ({ tasks }) => {
      const formatTaskDuration = jest
        .spyOn(utils, 'formatTaskDuration')
        .mockImplementation(() => 'date');

      const { getByTestId, getByText, queryByText } = renderTimeline(tasks);

      expect(queryByText('date')).not.toBeInTheDocument();

      fireEvent.mouseEnter(getByTestId('timeline-bar'));
      let tooltipExists = true;
      try {
        await waitFor(() => expect(getByText('date')).toBeInTheDocument());
      } catch {
        tooltipExists = false;
      }

      expect(tooltipExists).toBeFalsy();
      expect(formatTaskDuration).not.toBeCalled();
    }
  );
});
