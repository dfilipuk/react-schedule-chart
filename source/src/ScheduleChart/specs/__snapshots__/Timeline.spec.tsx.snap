// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Component Snapshot should not render current day marker if current day not within timeline 1`] = `
.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`Timeline Component Snapshot should render current day marker if current day within timeline 1`] = `
.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

.c1 {
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 9 / span 1;
  border-right: 1px dashed #606d7f;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
  </div>
</div>
`;

exports[`Timeline Component Snapshot should render several groups of several intersecting tasks 1`] = `
.c1 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: Gold;
  grid-area: 1 / 1 / 2 / 6;
}

.c4 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: LimeGreen;
  grid-area: 1 / 7 / 2 / 13;
}

.c2 {
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 2 / span 1;
  background-color: GoldenRod;
}

.c5 {
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 9 / span 1;
  background-color: ForestGreen;
}

.c3 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 3 / span 1;
}

.c6 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 11 / span 1;
}

.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="timeline-bar"
    />
    <div
      class="c2"
      data-testid="timeline-start-marker"
    />
    <div
      class="c3"
      data-testid="timeline-end-marker"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check-circle fa-w-16 fa-sm "
        color="GoldenRod"
        data-icon="check-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="c4"
      data-testid="timeline-bar"
    />
    <div
      class="c5"
      data-testid="timeline-start-marker"
    />
    <div
      class="c6"
      data-testid="timeline-end-marker"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check-circle fa-w-16 fa-sm "
        color="ForestGreen"
        data-icon="check-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Timeline Component Snapshot should render several not intersecting tasks 1`] = `
.c1 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: MediumVioletRed;
  grid-area: 1 / 1 / 2 / 4;
}

.c3 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: Gold;
  grid-area: 1 / 4 / 2 / 8;
}

.c4 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: LimeGreen;
  grid-area: 1 / 10 / 2 / 13;
}

.c2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 3 / span 1;
}

.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="timeline-bar"
    />
    <div
      class="c2"
      data-testid="timeline-end-marker"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check-circle fa-w-16 fa-sm "
        color="Purple"
        data-icon="check-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="c3"
      data-testid="timeline-bar"
    />
    <div
      class="c4"
      data-testid="timeline-bar"
    />
  </div>
</div>
`;

exports[`Timeline Component Snapshot should render several tasks which complete on single day 1`] = `
.c1 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: MediumVioletRed;
  grid-area: 1 / 5 / 2 / 13;
}

.c2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 12 / span 1;
}

.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="timeline-bar"
    />
    <div
      class="c2"
      data-testid="timeline-end-marker"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check-circle fa-w-16 fa-sm "
        color="Purple"
        data-icon="check-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Timeline Component Snapshot should render several tasks which start and complete on single day 1`] = `
.c1 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: Gold;
  grid-area: 1 / 3 / 2 / 12;
}

.c2 {
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 8 / span 1;
  background-color: GoldenRod;
}

.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="timeline-bar"
    />
    <div
      class="c2"
      data-testid="timeline-start-marker"
    />
  </div>
</div>
`;

exports[`Timeline Component Snapshot should render several tasks which start on single day 1`] = `
.c1 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: Gold;
  grid-area: 1 / 5 / 2 / 13;
}

.c2 {
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 5 / span 1;
  background-color: GoldenRod;
}

.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="timeline-bar"
    />
    <div
      class="c2"
      data-testid="timeline-start-marker"
    />
  </div>
</div>
`;

exports[`Timeline Component Snapshot should render single active multiday task 1`] = `
.c1 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: LimeGreen;
  grid-area: 1 / 3 / 2 / 11;
}

.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="timeline-bar"
    />
  </div>
</div>
`;

exports[`Timeline Component Snapshot should render single active one day task 1`] = `
.c1 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: Gold;
  grid-area: 1 / 7 / 2 / 8;
}

.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="timeline-bar"
    />
  </div>
</div>
`;

exports[`Timeline Component Snapshot should render single completed multiday task 1`] = `
.c1 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: MediumVioletRed;
  grid-area: 1 / 10 / 2 / 15;
}

.c2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 14 / span 1;
}

.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="timeline-bar"
    />
    <div
      class="c2"
      data-testid="timeline-end-marker"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check-circle fa-w-16 fa-sm "
        color="Purple"
        data-icon="check-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Timeline Component Snapshot should render single completed one day task 1`] = `
.c1 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: MediumVioletRed;
  grid-area: 1 / 1 / 2 / 2;
}

.c2 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 1 / span 1;
}

.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="timeline-bar"
    />
    <div
      class="c2"
      data-testid="timeline-end-marker"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check-circle fa-w-16 fa-sm "
        color="Purple"
        data-icon="check-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Timeline Component Snapshot should render single group of several intersecting tasks 1`] = `
.c1 {
  margin: 10px 2px;
  border-radius: 4px;
  background-color: Gold;
  grid-area: 1 / 1 / 2 / 15;
}

.c2 {
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 1 / span 1;
  background-color: GoldenRod;
}

.c4 {
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 4 / span 1;
  background-color: GoldenRod;
}

.c5 {
  width: 12px;
  height: 12px;
  border-radius: 50%;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 10 / span 1;
  background-color: GoldenRod;
}

.c3 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  justify-self: center;
  grid-row: 1 / span 1;
  grid-column: 3 / span 1;
}

.c0 {
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: repeat(14,1fr);
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="timeline-bar"
    />
    <div
      class="c2"
      data-testid="timeline-start-marker"
    />
    <div
      class="c3"
      data-testid="timeline-end-marker"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check-circle fa-w-16 fa-sm "
        color="GoldenRod"
        data-icon="check-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="c4"
      data-testid="timeline-start-marker"
    />
    <div
      class="c5"
      data-testid="timeline-start-marker"
    />
  </div>
</div>
`;
